design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,240 @@
+specification. The parent Orchestrator Module in Ceph will not store user-specified Drive Groups;
+however, it does expect the underlying Orchestrator or Orchestrator Mgr Module to store the Drive
+Group spec.
+
+Even with Drive Groups, Ceph expects the on-disk content of disks to be the declarative state of
+whether an OSD belongs to a Ceph cluster or not and what its configuration is. Rook is currently
+able to read this on-disk content via `ceph-volume` to determine if it should run a Kubernetes
+Deployment for a disk-based OSD.
+
+Rook currently does not implement OSD removal via any declarations in the `CephCluster` resource in
+order to prioritize the safety of user data. If a user wishes to remove an OSD, they must perform
+manual, imperative steps to do so.
+
+
+## Proposal
+
+### Feature requests to Ceph project
+This proposal will require changes to Rook and to Ceph's Rook Orchestrator Mgr Module. These changes
+are assumed, but requested changes to other Ceph components will be noted here.
+
+1. `ceph-volume` - Accept Drive Groups as input via command line flag
+   1. Should be automatically converted and issued `ceph-volume lvm batch` commands under the hood
+   2. Support JSON and YAML strings as inputs to command line flag
+   3. `ceph-volume` can ignore the node matcher glob patter, as Rook will handle node selection
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section to allow disks to be added via a Drive Groups specification. The spec is proposed below with
+behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage: # preexisting
+    driveGroups: < YAML blob of Drive Groups > # proposed new
+      # # Examples below
+      # regular_nodes:
+      #   host_pattern: '*-osd'
+      #   data_devices:
+      #     all: true
+      # fast_nodes:
+      #   host_pattern: '*-osd-fast'
+      #   data_devices:
+      #     limit: 6
+      #     size: "10TB:10TB"
+```
+
+`driveGroups` accepts a YAML blob of Drive Group specifications. Each Drive Group spec begins with a
+named key which is the name of the Drive Group, used to uniquely identify the Drive Group.
+
+`ceph-volume` feature request #1 will be used here to allow Rook to pass along the Drive Group YAML
+blob without modification or inspection to `ceph-volume` when running the OSD provisioning job. OSD
+creation via Drive Groups is given priority and executed before any other CRD-based OSD creation
+methods are used (the other CRD-based creations methods are still used after the Drive Group
+method).
+
+Drive Groups should be defined only at the cluster level, not at node level. Drive Groups support a
+standard glob matching pattern for nodes, and Rook should only run Drive Groups on the nodes which
+match this pattern. (As an aside, this also sets a foundational precedent for handling future
```

Comment:
If we go ahead and run it on all nodes, c-v will ignore the request when running on nodes that don't match, right? So we're just talking about an efficiency gain by not running it on all nodes. 

Currently every rook orchestration already runs the osd prepare job on all nodes (that match the placement, which by default is all nodes). The jobs are all executed in parallel so we're not talking about a delay for the orchestration to do this.

Running it on all nodes seems like a safer option than re-implementing the glob matching and risking missing nodes in the orchestration.

Reply:
I tried to bring this same point up in the meeting Friday, but @jan--f said that getting the host name in Python (i.e., `ceph-volume`) is really difficult. I think @sebastian-philipp was concerned about the efficiency.

Reply:
If we pass a hostname flag to c-v, this would resolve the complexity, right? Rook knows the host name.